### PR TITLE
AST: Fix compareDependentTypes() for protocol typealiases 

### PR DESCRIPTION
**Pull request by slavapestov**
Fixes <https://bugs.swift.org/browse/SR-3687> and <rdar://problem/30118513>.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: lib/AST/Type.cpp [hunk 1]
```diff
@@ -1221,6 +1221,7 @@ CanType TypeBase::getCanonicalType() {
 
     Result = GenericFunctionType::get(sig, inputTy, resultTy,
                                       function->getExtInfo());
+//    assert(function->hasError() || !Result->hasError());
```
Comment (DougGregor) [inline lib/AST/Type.cpp]: Did you mean to remove this?

**Review by DougGregor — APPROVED** [review]
Thanks!

File: lib/AST/Type.cpp (hunk 1 above)
Comment (slavapestov) [inline lib/AST/Type.cpp]: Yeah

Comment (slavapestov) [inline lib/AST/Type.cpp]: It looks like a nice assert but it made some compiler_crashers regress... so I'll look into this later.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Pull request merged**